photutils/detection/core.py

```diff
@@ -295,39 +314,112 @@ def find_peaks(data, threshold, min_separation=2, exclude_border=True,
+        `~astropy.table.Table`.
 
     Returns
     -------
     output : `~astropy.table.Table`
         A table containing the x and y pixel location of the peaks and
-        their values.
+        their values.  If ``subpixel=True``, then the table will also
+        contain the local centroid and fitted peak value.
     """
 
-    if segment_image is not None:
-        if segment_image.shape != data.shape:
-            raise ValueError('segment_image and data must have the same '
-                             'shape')
-
-    from skimage.feature import peak_local_max
-    coords = peak_local_max(data, min_distance=int(min_separation),
-                            threshold_abs=threshold, threshold_rel=0.0,
-                            exclude_border=exclude_border, indices=True,
-                            num_peaks=npeaks, footprint=footprint,
-                            labels=segment_image)
-    y_peaks, x_peaks = coords[:, 0], coords[:, 1]
+    from scipy import ndimage
+
+    if np.all(data == data.flat[0]):
+        return []
+
+    if footprint is not None:
+        data_max = ndimage.maximum_filter(data, footprint=footprint,
+                                          mode='constant', cval=0.0)
+    else:
+        data_max = ndimage.maximum_filter(data, size=box_size,
+                                          mode='constant', cval=0.0)
+
+    peak_goodmask = (data == data_max)    # good pixels are True
+
+    if mask is not None:
+        mask = np.asanyarray(mask)
+        if data.shape != mask.shape:
+            raise ValueError('data and mask must have the same shape')
+        peak_goodmask = np.logical_and(peak_goodmask, ~mask)
+
+    if border_width is not None:
+        for i in range(peak_goodmask.ndim):
+            peak_goodmask = peak_goodmask.swapaxes(0, i)
+            peak_goodmask[:border_width] = False
+            peak_goodmask[-border_width:] = False
+            peak_goodmask = peak_goodmask.swapaxes(0, i)
+
+    peak_goodmask = np.logical_and(peak_goodmask, (data > threshold))
+    y_peaks, x_peaks = peak_goodmask.nonzero()
     peak_values = data[y_peaks, x_peaks]
 
-    if coords.shape[0] > npeaks:
-        # NOTE: num_peaks is ignored by peak_local_max() if labels are input
+    if len(x_peaks) > npeaks:
         idx = np.argsort(peak_values)[::-1][:npeaks]
         x_peaks = x_peaks[idx]
         y_peaks = y_peaks[idx]
         peak_values = peak_values[idx]
 
-    return Table((x_peaks, y_peaks, peak_values),
-                 names=('x_peak', 'y_peak', 'peak_value'))
+    if subpixel:
+        x_centroid, y_centroid = [], []
+        fit_peak_values = []
+        for (y_peak, x_peak) in zip(y_peaks, x_peaks):
+            region = cutout_footprint(data, (x_peak, y_peak),
+                                      box_size=box_size, footprint=footprint,
+                                      mask=mask, error=error)
+            gaussian_fit = fit_2dgaussian(region[0], mask=region[1],
```

Comment:
IMO if you return more than one or two things it's better to return a bunch or dict or `namedtuple`, i.e. give the things you return names.
This makes for more readable code when you use some of the returned results ... in this case I have to look at the docs to be sure what [0], [1] and [2] are.
But I know that others prefer to just return long tuples ... so if you prefer this, OK!